crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,384 @@
+        .unwrap()
+});
+
+static ISSUE_LINK_REGEX_SET: Lazy<RegexSet> = Lazy::new(|| {
+    RegexSet::new([
+        r#"^#\s*(http|https)://.*"#, // issue link
+        r#"^#\s*\d+$"#,              // issue code - like "003"
+        r#"^#\s*[A-Z]{1,6}\-?\d+$"#, // issue code - like "TDO-003"
+    ])
+    .unwrap()
+});
+
+static NUM_CAPTURE_GROUPS: usize = 5usize;
+static TODO_LENGTH: usize = 4usize;
+
+pub fn check_todos(
+    tokens: &[LexResult],
+    autofix: flags::Autofix,
+    settings: &Settings,
+) -> Vec<Diagnostic> {
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+    let mut prev_token_is_todo = false;
+    let mut prev_token_todo_start = 2; // Default to 2, the position of "T" in a properly formed TODO
+
+    for (start, token, end) in tokens.iter().flatten() {
+        let diagnostics_ref = &mut diagnostics;
+        let range = Range::new(*start, *end);
+
+        let token_opt = match token {
+            Tok::Comment(s) => Some(s),
+            _ => None,
+        };
+
+        // Check for errors due to a missing link: TDO003.
+        if prev_token_is_todo {
+            if token_opt.is_some() && ISSUE_LINK_REGEX_SET.is_match(token_opt.unwrap()) {
+                prev_token_is_todo = false;
+                continue;
+            }
+
+            diagnostics_ref.push(Diagnostic::new(
+                MissingLinkInTodo,
+                Range::new(
+                    Location::new(start.row() - 1, prev_token_todo_start),
+                    Location::new(end.row() - 1, prev_token_todo_start + TODO_LENGTH),
+                ),
+            ));
+        }
+
+        let Some(comment) = token_opt else {
+            prev_token_is_todo = false;
+            continue;
+        };
+
+        let mut captures_opt = get_captured_matches(comment);
+        if captures_opt.peek().is_none() {
+            // If we didn't match the regex at all, we know that this token isn't a TODO. The regex
+            // defined above requires that the `tag` capture group is matched.
+            prev_token_is_todo = false;
+            continue;
+        };
+
+        // Check for errors on the tag: TDO001/TDO006.
+        // Unwrap is safe because the "tag" capture group is required to get here.
+        let captures = captures_opt.peek().unwrap();
+        let tag = captures.name("tag").unwrap().as_str();
+        if tag != "TODO" {
+            diagnostics_ref.push(Diagnostic::new(
+                InvalidTodoTag {
+                    tag: String::from(tag),
+                },
+                range,
+            ));
+
+            if tag.to_uppercase() == "TODO" {
+                let invalid_capitalization = Diagnostic::new(
+                    InvalidCapitalizationInTodo {
+                        tag: String::from(tag),
+                    },
+                    range,
+                )
+                .with_fix(
+                    if should_autofix(autofix, settings, Rule::InvalidCapitalizationInTodo) {
```

Comment:
nit: Should we use the same pattern as used at other places? Define a mutable diagnostic, check if autofix is enabled and then update the diagnostic with the edits.